src/utils/graphql/common.ts

```diff
@@ -0,0 +1,141 @@
+
+// TODO: add ao contract tags
+
+export const SMARTWEAVE_CONTRACT_TAGS = [
+  {
+    name: 'App-Name',
+    values: ['SmartWeaveContract'],
+  },
+];
+
+/**
+ * @param arweave - arweave instance to perform gql request
+ * @param address - address of the wallet to get transactions for
+ * @param protocolTags - the data protocol to filter by
+ * @param signal - signal to abort the request
+ * @returns @type {Promise<DataProtocolTransaction[]>} - returns the list of data protocol txs, for e.g smartweave contracts
+ */
+export async function getDataProtocolTransactionsByOwner({
+  address,
+  arweave,
+  protocolTags,
+  signal,
+}: {
+  arweave: Arweave;
+  address: string;
+  protocolTags: { name: string; values: string[] }[];
+  signal?: AbortSignal;
+}): Promise<DataProtocolTransaction[]> {
+  let hasNextPage = false;
+  let cursor: string | undefined;
+  const protocolTxs = new Set<DataProtocolTransaction>();
+  do {
+    const queryObject = {
+      query: `
+      { 
+          transactions (
+              owners:["${address}"]
+              tags: ${protocolTags},
+              sort: HEIGHT_DESC,
+              first: ${MAX_REQUEST_SIZE},
+              bundledIn: null,
+              ${cursor !== undefined ? `after: "${cursor}"` : ''}
+          ) {
+              pageInfo {
+                  hasNextPage
+              }
+              edges {
+                  cursor
+                  node {
+                      id
+                      block {
+                          height
+                      }
+                      tags {
+                          name
+                          value
+                      }
+                  }
+              }
+          }
+      }`,
+    };
+
+    if (signal?.aborted) {
+      throw new EvaluationTimeoutError('GraphQL request was aborted.');
+    }
+
+    const { status, ...response } = await arweave.api.post(
+      '/graphql',
+      queryObject,
+      {
+        signal,
+      },
+    );
+
+    if (status == 429) {
+      // Rate limited, wait for 10 seconds and retry
+      await new Promise((resolve) => setTimeout(resolve, 10_000));
+      continue;
+    }
+
+    if (status !== 200) {
+      throw Error(
+        `Failed to fetch contracts for wallet. Status code: ${status}`,
+      );
+    }
+
+    if (!response.data.data?.transactions?.edges?.length) {
+      continue;
+    }
+    response.data.data.transactions.edges
+      .map((e: { node: GQLNodeInterface }) => ({
+        id: e.node.id,
+        tags: e.node.tags,
+        data: e.node.data,
+      }))
+      .forEach((c: { def: DataProtocolTransaction; cursor: string }) => {
+        protocolTxs.add(c.def);
+      });
+    cursor =
+      response.data.data.transactions.edges[MAX_REQUEST_SIZE - 1]?.cursor;
+    hasNextPage =
+      response.data.data.transactions.pageInfo?.hasNextPage ?? false;
+  } while (hasNextPage);
+
+  return [...protocolTxs];
+}
+
+export async function getSmartweaveContractsFromGQL(params: {
```

Comment:
let's just put this all into one function, i don't think we need to separate them. there is a lot we are assuming in the `getDataProtocolTransactionsByOwner`, e.g. `bundledIn: null` so it's not really giving us any abstraction